### PR TITLE
this fixes the unpacking of the si_addr member from siginfo_t for the…

### PR DESCRIPTION
**Pull request by dhuseby**
… bsd oses

apparently not all OSes are the same? weird.

### CODE REVIEW

File: src/libstd/sys/unix/stack_overflow.rs [hunk 1]
```diff
@@ -71,7 +75,39 @@ mod imp {
...
+        #[repr(C)]
+        struct siginfo_t {
+            a: [libc::c_int; 3], // si_signo, si_code, si_errno
+            b: libc::pid_t,      // si_pid
+            c: libc::uid_t,      // si_gid
+            d: libc::c_int,      // si_status
+            si_addr: *mut libc::c_void,
+        }
+
+        (*(info as *const siginfo_t)).si_addr as usize
```
Comment (brson) [inline src/libstd/sys/unix/stack_overflow.rs]: Does this cast mean the definition in the libc crate is incorrect?

Comment (alexcrichton) [inline src/libstd/sys/unix/stack_overflow.rs]: These typically arise because `siginfo_t` is a union in C which we can't express well in libc, so we end up doing something like:

```rust
pub struct siginfo_t {
    pub si_signo: ::c_int,
    pub si_errno: ::c_int,
    pub si_code: ::c_int,
    pub _pad: [::c_int; 29],
}
```

where it's up to you to cast `&info._pad` to whatever struct you want out of the union.

@dhuseby it looks like freebsd/dragonfly  already have `si_addr` in this structure, though?

```rust
pub struct siginfo_t {
    pub si_signo: ::c_int,
    pub si_errno: ::c_int,
    pub si_code: ::c_int,
    pub si_pid: ::pid_t,
    pub si_uid: ::uid_t,
    pub si_status: ::c_int,
    pub si_addr: *mut ::c_void,
    _pad: [::c_int; 12],
}
```

File: src/libstd/sys/unix/stack_overflow.rs [hunk 2]
```diff
@@ -60,7 +60,11 @@ mod imp {
     // This is initialized in init() and only read from after
     static mut PAGE_SIZE: usize = 0;
 
-    #[cfg(any(target_os = "linux", target_os = "android"))]
+    #[cfg(any(target_os = "linux",
+              target_os = "android",
+              target_os = "bitrig",
```
Comment (alexcrichton) [inline src/libstd/sys/unix/stack_overflow.rs]: It looks like bitrig already defines the `si_addr` field, how come this was necessary?

```rust
pub struct siginfo_t {
    pub si_signo: ::c_int,
    pub si_code: ::c_int,
    pub si_errno: ::c_int,
    pub si_addr: *mut ::c_void
}
```

File: src/libstd/sys/unix/stack_overflow.rs [hunk 3]
```diff
@@ -60,7 +60,11 @@ mod imp {
...
+              target_os = "openbsd",
```
Comment (alexcrichton) [inline src/libstd/sys/unix/stack_overflow.rs]: same for openbsd:

```rust
pub struct siginfo_t {
    pub si_signo: ::c_int,
    pub si_code: ::c_int,
    pub si_errno: ::c_int,
    pub si_addr: *mut ::c_char,
    __pad: [u8; 108],
}
```

**Comment by alexcrichton** [comment]
I think that while the `si_addr` field is standard across all OS-es the actual layout of `siginfo_t` varies quite a bit. On some oses (e.g. Linux) it's made up of a bunch of unions so we can't actually write down a representation in the `libc` crate.

It looks like, however, the only OS without a `si_addr` field already defined is netbsd, the freebsd/dragonfly/openbsd/bitrig changes may not be required?

**Comment by dhuseby** [comment]
@alexcrichton oh man, you're right about freebsd/bitrig.  I was working from an old libc revision.  Let's close this one out and let me try again.  I'll put as much in libc as I can.  Sorry for wasting your time.

**Comment by alexcrichton** [comment]
Oh no worries! It looks like the netbsd definition in libc may either need to be updated or support needs to be added here. I'm not sure if `siginfo_t` is a union on the BSDs like it is on Linux, it looked like on OSX at least it was just a plain struct which would make our lives much easier!

If it ends up being a struct, feel free to fill out the definitions!

**Comment by dhuseby** [comment]
@alexcrichton it's a really messy union on the BSD's and it seems like they all have their own version.

**Comment by alexcrichton** [comment]
Ah well, one could always hope...
